include/mvdtool/bits/mvd3_misc.hpp

```diff
@@ -95,37 +95,37 @@ std::vector<T> tsv_get_chunked(const MVD3::MVD3File& mvd,
 
 
 // cells properties
-const std::string did_cells_positions = "/cells/positions";
-const std::string did_cells_rotations = "/cells/orientations";
+constexpr char did_cells_positions[] = "/cells/positions";
```

Comment:
doesn't this mean that std::strings are implicitly created at each point of use, on each execution?

Reply:
If I read this right: https://godbolt.org/z/rn3qwC

Just means that the expression can be evaluated at compile time, and not at run time.

Reply:
So there does not seem to be duplication of data, if that's what you're worried about?

Reply:
I agree with @matz-e - this should be completely compile time and produce a string literal.

Reply:
No, I mean that, `did_cells_positions` is used like
```_hdf5_file.getDataSet(did_cells_positions)```

where `getDataSet` takes a std::string, so all these constants are getting having `std::strings` constructed out of them.

Reply:
ie: https://godbolt.org/z/pQcJN3

Reply:
I've played with this too long. I guess Python pins us to old-school C++. Would this be an acceptable penalty?

In a way, I feel like we're arguing about a few memory operations that are then followed by I/O on a parallel filesystem…

Reply:
I just wanted to point out what this entails: here it might be acceptable for performance, but it's not always the case.

I'm not a fan of global ctors, so I'd normally put them behind a getter that init's them once on use.  On the other hand, I'm not sure having `constexpr` is really what you want for things that are (only?) used as `std::strings`.

Ultimately it's up to you guys.

Reply:
@ohm314 your thoughts, pretty please?

IMO, especially when using this from Python, performance impact should be small, if not completely negligible. Once we've collectively moved to more modern tooling, we can update this to `constexpr std::string`. If there are no further objections, I'll merge this soon™.

Reply:
I'd add a comment why this is needed, or potentially a test that exhibits 'but no values would be assigned to the dataset ids and runtime exceptions will be thrown.' if it were changed back to something else.

Reply:
OK, I'll do that

Reply:
This definitely took too long and made me doubt my own sanity… I'm still not entirely sure what's going on when you take the new test and build it with the `const std::string` header. I guess the ordering of the string construction is coming later than the instantiation of the pseudo object I have there…

Reply:
Yeah, static ctors are evil, and should be avoided.  Your test isn't guaranteed to fail all the time, though, since it's at the whim of the linker/glibc.

Like I said, it seems the 'right' fix w/ the current language standard would be to wrap the std:strings in getters.  Such is the C++ world :(

Reply:
Yes, I was wondering if it's a linking order issue. Anyways… we have getters in synapse-tool, but they still construct a string every time called if I read the code correctly. Makes for more complex code in the end.

I think this is good enough here, a better solution IMO would be to bump the C++ standard *or* have HighFive accept both `std::string` and `const char*` for dataset names…

The static ctor is indeed not good, just seems like it should be possible to use things like that.

Reply:
> Anyways… we have getters in synapse-tool, but they still construct a string every time called if I read the code correctly.

IIRC, this is the pattern:
```
const std::string& did_cells_positions()
{
    static const std::string s = "/cells/positions";
    return s;
}
```

It's things like this why I'm not a huge fan of C++...

Reply:
this is ugly. What is so bad about?
```c++
std::string posics() {
    return "/cells/positions";
}
```
Since literals go into the data section, and C++ assumes string literals as std::string, and RVO applies here, doesnt the compiler translated this to almost perfect code? (something wrong in my thinking?)

Reply:
>  .. C++ assumes string literals as std::string ...

I don't think that is true; the literal is just a char array.

The `posics()` as you've written it will make a std::string on each call, with the `malloc` that is associated with that, and then copy (likely) `memcpy` "/cells/positions" each time.  The static way means that this allocation and copy only happens once.

Reply:
> What I fail to see is why the change to char[].

`constexpr std::string` is C++20…

Reply:
Oh yeah, and
> IIRC, this is the pattern:
then we go from one line of code in a header-only I/O library to 5…

Reply:
> then we go from one line of code in a header-only I/O library to 5…

Maybe I'm not reading this right, but I don't think that's the best metric for writing code, especially in a language that is notoriously verbose.

Also, I think the focus of this PR being constexpr is a bit of a red-herring.  The real problem was that there are global ctors in the code, and you can end falling into the order of initialization not being defined problem.  There are other places in the code that has this same pattern, and those should probably be fixed, too, before someone runs into them.

How to fix this is partly covered by the discussion above, but I'm not sure there has been a clear outcome from it.  AFAIK, the static string w/ a wrapper I gave before is the lowest overhead (except for line count), and I would be thrilled to know of a better way/pattern.

Reply:
What about using `string_view` instead of `string` on the other side?

Reply:
I know it's a red herring, and this solution may be lacking. The added benefit of the wrapping functions would be the disentanglement of headers and definitions in *compiled* libraries.

We could also add functions accepting `const char*` in HighFive to get around constructing a whole bunch of strings. And then inline the functions accepting `std::string`…

Reply:
> What about using string_view instead of string on the other side?

Do you mean in HighFive?

> We could also add functions accepting const char* in HighFive to get around constructing a whole bunch of strings.

I was hesitent to suggest this, but that thought also crossed my mind.  I had looked at the HighFive impl, and std::strings area used a lot.

Reply:
Why not in HighFive but I look, it's C++17, sorry for the noise... But it's the answer, if you don't want to modify the string coerce it to a view.

Reply:
> But it's the answer, if you don't want to modify the string coerce it to a view.

IIRC, the h5 wants a null terminated string, and string_view doesn't handle that transparently, I believe.